jwt/jwt.go

```diff
@@ -0,0 +1,202 @@
+
+	if initError != nil {
+		return nil, initError
+	}
+
+	if jwtUtilsInstance == nil {
+		return nil, errors.New("internal server error")
+	}
+
+	return jwtUtilsInstance, nil
+}
+
+func (j *JWTUtils) initialize() error {
+	if j == nil {
+		return errors.New("internal server error")
+	}
+
+	publicKeyString, err := getPublicKeyFromParameterStore("publickey")
+	if err != nil {
+		return err
+	}
+
+	block, _ := pem.Decode([]byte(publicKeyString))
+	if block == nil {
+		return errors.New("internal server error")
+	}
+
+	pub, err := x509.ParsePKIXPublicKey(block.Bytes)
+	if err != nil {
+		return fmt.Errorf("internal server error")
+	}
+
+	rsaPublicKey, ok := pub.(*rsa.PublicKey)
+	if !ok {
+		return errors.New("internal server error")
+	}
+
+	j.publicKey = rsaPublicKey
+	return nil
+}
+
+func getPublicKeyFromParameterStore(parameterName string) (string, error) {
+	sess, err := session.NewSession()
+	if err != nil {
+		return "", err
+	}
+
+	svc := ssm.New(sess)
+	input := &ssm.GetParameterInput{
+		Name:           aws.String(parameterName),
+		WithDecryption: aws.Bool(true),
+	}
+
+	result, err := svc.GetParameter(input)
+	if err != nil {
+		return "", err
+	}
+
+	return *result.Parameter.Value, nil
+}
+
+func (j *JWTUtils) ValidateToken(tokenString string) (jwt.MapClaims, error) {
+	if j == nil || j.publicKey == nil {
+		return nil, errors.New("internal server error")
+	}
+
+	token, err := jwt.Parse(tokenString, func(token *jwt.Token) (interface{}, error) {
+		if _, ok := token.Method.(*jwt.SigningMethodRSA); !ok {
+			return nil, fmt.Errorf("invalid token")
+		}
+		return j.publicKey, nil
+	})
+
+	if err != nil {
+		return nil, fmt.Errorf("invalid token")
+	}
+
+	if !token.Valid {
+		return nil, errors.New("invalid token")
+	}
+
+	claims, ok := token.Claims.(jwt.MapClaims)
+	if !ok {
+		return nil, errors.New("invalid token")
+	}
+
+	return claims, nil
+}
+
+func (j *JWTUtils) ExtractClaim(claims jwt.MapClaims, claimKey string) (string, error) {
+	if claims == nil {
+		return "", errors.New("internal server error")
+	}
+
+	value, ok := claims[claimKey].(string)
+	if !ok || value == "" {
+		return "", fmt.Errorf("unauthorized")
+	}
+	return value, nil
+}
+
+func JWTMiddleware() func(req events.APIGatewayProxyRequest) (events.APIGatewayProxyResponse, string, error) {
+	return func(req events.APIGatewayProxyRequest) (events.APIGatewayProxyResponse, string, error) {
+		jwtUtils, err := GetInstance()
+		if err != nil {
+			log.Printf("Failed to get JWTUtils instance: %v", err)
+			return events.APIGatewayProxyResponse{
+				StatusCode: http.StatusInternalServerError,
+				Body:       "Internal server error",
+			}, "", nil
+		}
+
+		cookie := req.Headers["Cookie"]
+		if cookie == "" {
+			return events.APIGatewayProxyResponse{
+				StatusCode: http.StatusUnauthorized,
+				Body:       "Unauthenticated",
+			}, "", nil
+		}
+
+		cookieName := os.Getenv("SESSION_COOKIE_NAME")
+		if cookieName == "" {
+			cookieName = utils.DEVELOPMENT_COOKIE_NAME
+
+		}
```

Comment:
- instead of setting the cookie name here can we move this to `utils/constant.go` ?
- instead of setting the `DEVELOPMENT_COOKIE_NAME` when `cookieName` is empty can we have an `env` flag on which we can perform a check similar to how we're doing in `template.yaml`

Reply:
done